### ISSUE TITLE
[Feature request] Text object for comments

### ISSUE THREAD

**Issue by ouuan** [numToStr/Comment.nvim#11]
Basically the same as https://github.com/tomtom/tcomment_vim/issues/237

**Comment by astier** [comment]
I think this should be three regions. I agree with the examples you made.

**Comment by RRethy** [comment]
https://github.com/RRethy/nvim-treesitter-textsubjects has single and multiline comment support in the `textsubjects-smart` query (although there is a bug in tree-sitter queries which makes it limited), another query could be added specifically for comments.

**Comment by numToStr** [comment]
@RRethy If that is possible then it would be great ;)

**Comment by ouuan** [comment]
I just discovered that https://github.com/nvim-treesitter/nvim-treesitter-textobjects also have `@comment.outer`.

**Comment by ouuan** [comment]
> I just discovered that [nvim-treesitter/nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects) also have `@comment.outer`.
> 
> (Also mentioned at [#77 (comment)](https://github.com/numToStr/Comment.nvim/issues/77#issuecomment-974798712))

Unfortunately, it doesn't work with line comments very well:

1. `@comment.outer` only matches a single line when there are multiple line comments in a row.
2. `@comment.outer` matches exactly the comment instead of the whole line, so `Comment.nvim` will turn a line comment as the content of a block comment. e.g. it turns `// comment` to `/* // comment */`.

**Comment by numToStr** [comment]
@ouuan Yes, it doesn't work with line comments. Although it works very well with block comments. 

Regarding the feature itself, I am still confused as to whether should I implement this myself or let the other plugins like https://github.com/nvim-treesitter/nvim-treesitter-textobjects or https://github.com/RRethy/nvim-treesitter-textsubjects provide the text objects. If any of the plugins can achieve this https://github.com/numToStr/Comment.nvim/issues/11#issuecomment-944922532 behavior then I would recommend using that plugin.

**Comment by ouuan** [comment]
I agree that Comment.nvim doesn't need to implement this feature. It could be a part of another plugin or a new plugin.

**Comment by ouuan** [comment]
There's also a problem with multiline block comments.

`Comment.nvim` adds `//` at the beginning of each line:

```cpp
/**
 * multiline comment
 */
```

**Comment by numToStr** [comment]
I think you are using the `gc` bindings. Try with `gb`.

**Comment by numToStr** [comment]
@ouuan Also make sure that the plugin contains queries for comments. The last time I checked https://github.com/nvim-treesitter/nvim-treesitter-textobjects was missing comments queries for some languages.

**Comment by clason** [comment]
Yes, and help is welcome :)

**Comment by RRethy** [comment]
There's a bug in tree-sitter queries which is blocking on this, you can see it with this query:

```scm
(((_) @head . (comment) @_start . (comment)+ @_end (_) @tail)
    (#not-has-type? @tail "comment")
    (#not-has-type? @head "comment")
    (#make-range! "range" @_start @_end))
```

It will find a match in this file:

```ruby
def foo
  # foo
  # foo
  # foo
  puts 'hello'
end
```

But it will fail to find a match in this file with more comments, even though there should be a match.

```ruby
def foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  # foo
  puts 'hello'
end
```

I haven't filed anything yet, just been busy.

**Comment by Ciel-MC** [comment]
mini.comment already has this, perhaps inspirations can be drawn there? I'm not aware of the architecture but this would be really helpful, since as of now I have to choose between block comments or comment text objects, or make some Chimera of the two, none of which is ideal.
